feat: enable passing generation_kwargs to the PromptNode in pipeline.run()

faaany opened the pull request:
### Related Issues
- fixes n/a

### Proposed Changes:
- add a new optional parameter `generation_kwargs` in the run method of PromptNode object 
- update the `invocation_context` variable with the additional params defined in `generation_kwargs`
For more info related to `generation_kwargs`, pls refer to [this doc](https://huggingface.co/docs/transformers/main/en/generation_strategies#customize-text-generation).
- enable `max_new_tokens` customization 

### How did you test it?
Locally, I tried 2 cases. And the generated texts are different, indicating that the params are passed sucessfully.

```python
result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
    "Prompter": { "prompt_template": "question-answering"},
    "Retriever": {"top_k": 1},
    "Reranker": {"top_k": 1}
    })
```

```python
result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
    "Prompter": { "prompt_template": "question-answering", "generation_kwargs": { "do_sample": True}},
    "Retriever": {"top_k": 1},
    "Reranker": {"top_k": 1}
    })
```

### Notes for the reviewer
n/a

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [ ] I have updated the related issue with new insights and changes
- [x] I added tests that demonstrate the correct behavior of the change
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

vblagoje commented:
Great stuff @Faaany - we need this. Please create one unit test showing how we use generation_kwargs in a component's constructor and then override those values in the pipeline run method. Use some simple mocking to show that default generation_kwargs were indeed overridden in the call.

Faaany commented:
@vblagoje Sure, will do!

Faaany commented:
@vblagoje I added the unit test. Could you pls help to review the code? Thanks!

Faaany commented:
@vblagoje forgot to format the test_promode_node.py as well. Fixed with the new commit. Thanks for triggering the tests!

vblagoje commented:
@Faaany I had a closer look at this PR now. Can you please add a mock test where we demonstrate how generation_kwargs get passed as pipeline params and then get resolved in the prompt node, just as you demonstrated in your example:
```
result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
    "Prompter": { "prompt_template": "question-answering", "generation_kwargs": { "do_sample": True}},
    "Retriever": {"top_k": 1},
    "Reranker": {"top_k": 1}
    })
```

While this is a much-needed feature for HF-based PromptNodes, what about other providers? Can we generalize your approach to add support for all other providers as well? Perhaps we can unpack generation_kwargs and pass them to the underlying invocation layer via invoke method kwargs? That would be an amazing feature. Let me know your thoughts.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/deepset-ai/haystack?pullRequest=4832) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/deepset-ai/haystack?pullRequest=4832) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/deepset-ai/haystack?pullRequest=4832) it.</sub>

Faaany commented:
@vblagoje I added a new unit test for the pipeline call. Please have a look.

That's a good idea! But can we make a separate PR for the other providers? Since I don't have open_ai tokens nor azure account, it may take a bit of time for me to have both tested.

Faaany commented:
@vblagoje can we merge the PR? or is there anything I need to update?

vblagoje commented:
> @vblagoje can we merge the PR? or is there anything I need to update?

@Faaany good news, we think we should use the generation_kwargs parameter for all PromptNode invocation layers. After all these are key/value pairs in a dictionary and we can apply them to Cohere, OpenAI, Anthropic etc.

1. Would you please remove docs references where generation_kwargs is referred to be a HF specific parameter?

2. Please write another test showing how these generation_kwargs are passed to OpenAIInvocationLayer invoke for example if user initialized PromptNode with say gpt-3.5-turbo model. 

3. I or you can update unpacking of generation kwargs into kwargs passed to each model. Let's do that perhaps in each layer invoke impl.

Let me know your thoughts.

Faaany commented:
@vblagoje I have done the following and updated my PR:

- removed the docs references where generation_kwargs is referred to be a HF specific param
- updated the unit test; instead of writing a new one, I used the `pytest.mark.parametrize` to test different models. Hope it works 
- updated the invoke methods in each provider. Please let me know If I haven't covered all providers.

Thanks for the review!

vblagoje commented on test/prompt/test_prompt_node.py:
```diff
@@ -250,6 +250,23 @@ def test_generation_kwargs_from_prompt_node_call():
 
 
+@pytest.mark.integration
+@pytest.mark.parametrize("prompt_model", ["google/flan-t5-base", "gpt-3.5-turbo"], indirect=True)
```
@Faaany Let's check node.prompt_model.model_invocation_layer.pipe for HF only. For other models we can check generic invoke. Something like:

```
with patch.object(node.prompt_model.model_invocation_layer, "invoke", MagicMock())
...
...
```

Faaany replied:
How about this?

```python
@pytest.mark.integration
@pytest.mark.parametrize("prompt_model", ["hf", "openai", "azure"], indirect=True)
def test_generation_kwargs_from_prompt_node_run_hf(prompt_model):
    skip_test_for_invalid_key(prompt_model)
    the_question = "What does 42 mean?"
    # test that generation_kwargs are passed to the underlying invocation layer
    node = PromptNode(prompt_model)
    with patch.object(node.prompt_model.model_invocation_layer, "invoke", MagicMock()) as mock_call:
        node.run(query=the_question, prompt_template="{query}", generation_kwargs={"do_sample": True, "test": True})

        mock_call.assert_called_with(
            prompt=the_question,
            stop_words=None,
            top_k=1,
            query=the_question,
            generation_kwargs={"do_sample": True, "test": True}
        )
```

vblagoje replied:
Yes, great @Faaany  that's exactly what I had on mind as well

vblagoje commented on test/prompt/test_prompt_node.py:
```diff
@@ -724,6 +741,37 @@ def test_simple_pipeline_yaml(tmp_path):
+    the_question = "What is the capital of Germany?"
+    pipeline = Pipeline.load_from_yaml(path=tmp_path / "tmp_config.yml")
+
+    with patch.object(HFLocalInvocationLayer, "invoke", MagicMock()) as mock_call:
```
Yes, something like this mocking approach should work for all layers.

Faaany replied:
the problem with all other layers is that they require `api_key`.

vblagoje reviewed (changes requested):
Almost there; let's just improve unit tests a bit; see comments.

Faaany commented:
@vblagoje can we merge now?

vblagoje commented:
@Faaany, no problem, and congratulations! Let's wait for CI to finish, and I'd like to test this out locally once again, and we can proceed with PR approval.

silvanocerza commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -358,6 +360,9 @@ def run(
         if "prompt_template" not in invocation_context.keys():
             invocation_context["prompt_template"] = self.get_prompt_template(prompt_template)
 
+        if generation_kwargs and "generation_kwargs" not in invocation_context.keys():
+            invocation_context["generation_kwargs"] = generation_kwargs
```
I think it would be better if we actually merged this two. The `invocation_context` is passed as kwargs to `PromptNode.prompt_model.invoke()` in the end.

Making this change wouldn't require us to modify any existing and future invocation layer to handle custom kwargs, it's much more maintanable.

Faaany replied:
from the result and code simplicity point of view, the 2 params can indeed be merged; however, the meaning of these two parameters are different, `invocation_context` refers to all the parameters passed to the invoke function of the following layers, and `generation_kwargs` refers to the parameters that affect the performance of the model generation alone. Therefore, `generation_kwargs` should be a subset of `invocation_context`. Furthermore, `invocation_context` is the output of the Shaper, if there is a Shaper node in the pipeline. In this case, the invocation_context can't be passed as an argument, e.g. for this code snippet:

```python
result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
    "Prompter": { "prompt_template": "question-answering", "invocation_context": {"generation_kwargs": {"do_sample": True}}},
    "Retriever": {"top_k": 1},
    "Reranker": {"top_k": 1}
    })
```

Below is the error log for 
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 552, in run
    node_output, stream_id = self._run_node(node_id, node_input)
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 467, in _run_node
    return self.graph.nodes[node_id]["component"]._dispatch_run(**node_input)
  File "/usr/local/lib/python3.8/dist-packages/haystack/nodes/base.py", line 201, in _dispatch_run
    return self._dispatch_run_general(self.run, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/haystack/nodes/base.py", line 245, in _dispatch_run_general
    output, stream = run_method(**run_inputs, **run_params)
TypeError: run() got multiple values for keyword argument 'invocation_context'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "run_pipeline.py", line 18, in <module>
    result = pipeline.run(query="Why is Q1 guide revenue down so much vs what said previously?", params={
  File "/usr/local/lib/python3.8/dist-packages/haystack/pipelines/base.py", line 559, in run
    raise Exception(
Exception: Exception while running node 'Prompter': run() got multiple values for keyword argument 'invocation_context'
Enable debug logging to see the data that was passed when the pipeline failed.
```

silvanocerza replied:
Oh no, I didn't mean to change the naming of the `PromptNode` arg. What I mean is change the name of the param to `invocation_kwargs` as I said in the other comment and here instead of adding it to `invocation_context` as a new field we merge them together. 

Something like this.

```suggestion
        invocation_context.update(invocation_kwargs)
```

This way we don't need to change anything else, nor in the `PromptNode` nor in any invocation layer.

silvanocerza replied:
I didn't give it a try to be fair, but I'd expect this to work.

Faaany replied:
I see your point. But this approach would be different from what @vblagoje suggested earlier. Can we have a quick chat in Discord? I have sent you guys friend requests. That will be more efficient.

silvanocerza replied:
Ok, I'm ok with the naming. 

But what's stopping us from just merging `generation_kwargs` with `invocation_context` if in the end we just merge it with `invoke`'s `kwargs`?

The `invocation_context` is exploded as `PromptNode.__call__()`'s `kwargs` in line [366](https://github.com/deepset-ai/haystack/pull/4832/files#diff-977894d453c60a383b8c831ae90d32076823f4082a3ea5e9792054b33299929eR366), then again as `kwargs` to `PromptNode.prompt()` and then end up in [`PromptModel.invoke` as `kwargs_copy`](https://github.com/deepset-ai/haystack/blob/1dfb7454bfbe20a7c6f6947163cd7955725ea89b/haystack/nodes/prompt/prompt_node.py#LL174C68-L174C68) that just proxies them to the `InvocationLayer` used.

vblagoje replied:
Yeah by all means - if possible why not!

faaany replied:
> Ok, I'm ok with the naming.
> 
> But what's stopping us from just merging `generation_kwargs` with `invocation_context` if in the end we just merge it with `invoke`'s `kwargs`?
> 
> The `invocation_context` is exploded as `PromptNode.__call__()`'s `kwargs` in line [366](https://github.com/deepset-ai/haystack/pull/4832/files#diff-977894d453c60a383b8c831ae90d32076823f4082a3ea5e9792054b33299929eR366), then again as `kwargs` to `PromptNode.prompt()` and then end up in [`PromptModel.invoke` as `kwargs_copy`](https://github.com/deepset-ai/haystack/blob/1dfb7454bfbe20a7c6f6947163cd7955725ea89b/haystack/nodes/prompt/prompt_node.py#LL174C68-L174C68) that just proxies them to the `InvocationLayer` used.

the only problem I see with this approach is that we need to list all possible params of generation_kwargs in the line 183 of hugging_face.py. But compared to the fact that we have to update all providers' invocation layers, this approach has much less code change and is easier to maintain. So I updated the PR, pls review it. Thanks a lot!

silvanocerza reviewed (changes requested):
Left a comment that I think would make the code simpler. 

Also is the arg name clear enough? Wouldn't `invocation_kwargs` or something similar could be clearer? Not really sure about this though.

Faaany commented:
> Left a comment that I think would make the code simpler.
> 
> Also is the arg name clear enough? Wouldn't `invocation_kwargs` or something similar could be clearer? Not really sure about this though.

agree to use `invocation_kwargs` instead of `invocation_context`, because `invocation_context` is passed to `kwargs` in the following functions anyway and except the `run` method which used the wording `context`, all other functions used `kwargs`.

silvanocerza reviewed (approved):
Nice job simplify the code after the feedback, looking forward to future contributions. :)
